amlb/utils/process.py

```diff
@@ -186,6 +210,9 @@ def read_pipe(pipe, timeout):
         pipes = as_list(pipe)
         # wait until a pipe is ready for reading, non-Windows only.
         ready, *_ = select.select(pipes, [], [], timeout)
+        # if a pipe is not ready it could be timeout or it could be end of process
+        # so at this point we do not know. Only after the communicate function is over do we know.
+        # i.e., if the process is still running it does not have a retcode.
```

Comment:
Preferably I would have raised the activity timeout from the function which uses it. But at this stage we can unfortunately not detect whether the error should be raised.